New recaptcha solving method

gcosta opened the pull request:
As discussed earlier, this PR adds a task to solve recaptchas by hand. Let me explain the implementation details below.

First we added a new action type called "recaptcha". This step is used as a way to identify forms who require recaptcha solving and also to perform different actions depending on the execution context. For example, when filling out a form with poltergeist or webkit, the recaptcha action fails the fill instantly so it can go to the failed queue as soon as possible. When filling out a form with watir, the recaptcha step will sleep a few seconds to avoid a watir timeout and further actions while the user solves the recaptcha and submits the form by hand. Having a new action type also ensures that contact-congress implementations other than phantom-dc can safely ignore the step and solve the recaptcha their own way.

Second, the rake task itself. To avoid task duplication, we made minor edits to the main perform_fills task to add the recaptcha fills mode and to make it ignore recaptcha jobs by default to save time. Then we added a new task called perform_recaptcha_fills that "inherits" the perform_fills task and runs it with recaptcha mode activated while accepting the same params.

In recaptcha fills mode only recaptcha jobs will be run, and the fill out method is overridden to use the watir method, regardless of the form settings.

@j-ro  has been solving our own recaptcha forms with it and can describe the resulting workflow better.

j-ro commented:
workflow is pretty good! After getting set up locally and connecting with the production db, you run the task while you have a server running and that's kind of it.

Firefox pops up, the form pre-fills with the data, and you click the captcha and solve any challenges, then submit the form. Once it's submitted, switch over to the terminal and give the Control-C command to kill firefox and move on to the next one. New window, new form, solve captcha, submit, control-C, repeat!

j-ro commented:
@Hainish let us know when you've had a chance to look at this. It's working great for me! Ready to merge when you are.

Hainish commented on tasks/phantom-dc.rake:
```diff
@@ -24,15 +25,33 @@ namespace :'phantom-dc' do
           if regex.nil? or regex.match(cm.bioguide_id)
-            if retrieve_captchad_cached(captcha_hash, cm.id)
+            if cm.has_google_recaptcha?
+              recaptcha_jobs.push job
```
You probably want to do something like building a `recaptcha_hash` as we do for the `captcha_hash` above.

The reason we do this is to avoid an unnecessary database lookup of the member of congress associated with the current job, which happens for every job within this loop.  If there are some number of thousands of jobs, this results in the sieving step into `recaptcha_jobs`, `captcha_jobs`, and `noncaptcha_jobs` here to take much longer.

If you build a hash of the recaptcha'd members from the outset, there is only one lookup per MoC.  You could even build it into the `build_captcha_hash` method, to make line 22 above read:

```ruby
captcha_hash, recaptcha_hash = build_captcha_hash
```

Hainish commented:
@gcosta other than the one comment I have, lgtm.

gcosta commented:
Refactored it so the perform fills task will also build a hash for recaptcha. Also removed line 71 of the rakefile because the captcha hash it was building was never used there.
